src/kfactory/schema.py

```diff
@@ -0,0 +1,847 @@
+        return placement
+
+    @overload
+    def __getitem__(self, value: str) -> PortRef: ...
+    @overload
+    def __getitem__(self, value: tuple[str, int, int]) -> PortArrayRef: ...
+
+    def __getitem__(self, value: str | tuple[str, int, int]) -> PortRef:
+        if isinstance(value, str):
+            return PortRef(instance=self.name, port=value)
+        return PortArrayRef(instance=self.name, port=value[0], ia=value[1], ib=value[2])
+
+    def connect(
+        self, port: str | tuple[str, int, int], other: Port[TUnit] | PortRef
+    ) -> Connection[TUnit]:
+        if isinstance(port, str):
+            pref = PortRef(instance=self.name, port=port)
+        else:
+            pref = PortArrayRef(
+                instance=self.name, port=port[0], ia=port[1], ib=port[2]
+            )
+        conn = Connection[TUnit]((other, pref))
+        self.parent_schema.connections.append(conn)
+        return conn
+
+
+class Route(BaseModel, Generic[TUnit], extra="forbid"):
+    name: str = Field(exclude=True)
+    links: list[Link[TUnit]]
+    routing_strategy: str = "route_bundle"
+    settings: dict[str, JSONSerializable]
+
+    @model_validator(mode="before")
+    @classmethod
+    def _parse_links(cls, data: dict[str, Any]) -> dict[str, Any]:
+        links = data.get("links", [])
+
+        if isinstance(links, dict):
+            data["links"] = [
+                [tuple(k.split(",")), tuple(v.split(","))] for k, v in links.items()
+            ]
+        return data
+
+
+class Port(BaseModel, Generic[TUnit], extra="forbid"):
+    name: str = Field(exclude=True)
+    x: TUnit | PortRef | PortArrayRef
+    y: TUnit | PortRef | PortArrayRef
+    dx: TUnit = cast("TUnit", 0)
+    dy: TUnit = cast("TUnit", 0)
+    cross_section: str
+    orientation: Literal[0, 90, 180, 270]
+
+    def __lt__(self, other: Port[Any] | PortRef) -> bool:
+        if isinstance(other, Port):
+            return (
+                self.name,
+                self.x,
+                self.y,
+                self.dx,
+                self.dy,
+                self.orientation,
+                self.cross_section,
+            ) < (
+                other.name,
+                other.x,
+                other.y,
+                other.dx,
+                other.dy,
+                other.orientation,
+                other.cross_section,
+            )
+        return True
+
+    def is_placeable(self, c: ProtoTKCell[Any], placed_instances: set[str]) -> bool:
+        placeable = True
+        if isinstance(self.x, PortRef):
+            placeable = self.x.instance in placed_instances
+        if isinstance(self.y, PortRef):
+            placeable = placeable and self.y.instance in placed_instances
+        return placeable
+
+
+class Link(
+    RootModel[
+        tuple[
+            Port[TUnit] | PortArrayRef | PortRef, Port[TUnit] | PortArrayRef | PortRef
+        ]
+    ]
+):
+    root: tuple[PortArrayRef | PortRef, PortArrayRef | PortRef]
```

Comment:
Why do we overwrite this differently here?

Reply:
Good question. I guess forgot to change it. But since links are only used in routes, I would suggest to change it to `tuple[PortArrayRef | PortRef, PortArrayRef | PortRef]`
```suggestion
class Link(
    RootModel[
        tuple[
            PortArrayRef | PortRef, PortArrayRef | PortRef
        ]
    ]
):
    root: tuple[PortArrayRef | PortRef, PortArrayRef | PortRef]
```